### PR TITLE
fix: tool call message role according to credentials

### LINKED ISSUE #5622: Wrong message role when using openai_api_compatible llm in function calling agent mode
### Self Checks

- [X] This is only for bug report, if you would like to ask a question, please head to [Discussions](https://github.com/langgenius/dify/discussions/categories/general).
- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### Dify version

0.6.11

### Cloud or Self Hosted

Self Hosted (Source)

### Steps to reproduce

- When adding a new openai_api_compatible LLM, choose Tool Call for Function calling. Subsequently, during the creation of an Agent, I can utilize the Function calling as Agent mode.

截屏2024-06-26 15 14 03

- Then the agent uses tools and making tool calls without issue. However, after receiving the tool's response, attempting to invoke the large language model again results in a 400 error.

`[openai_api_compatible] Error: API request failed with status code 400: {"error":{"message":"bad response status code 400 (request id: 2024062614595215754623335617148)","type":"upstream_error","param":"400","code":"bad_response_status_code"}}`

- I inspected the logs of the deployment framework (which uses Xinference) providing the openai_api_compatible LLM service. The log indicated: `Exception: [address=0.0.0.0:33477, pid=1686] Unsupported message role: function`.

- According to my understanding and the [OpenAI API doc](https://platform.openai.com/docs/api-reference/chat/create#chat-create-messages), by selecting Tool Call as Function Calling method, the message role for the tool call's response should be `tool`. Conversely, if Function Call as Function Calling method is chosen, the message role should be `function`.

- After reviewing the code, I found that it does not differentiate and uniformly uses `function` as the tool message role. Many openai_api_compatible service frameworks, such as TGI and Xinference, employ `tool` as the tool message role. This discrepancy can led to the bug.

### ✔️ Expected Behavior

When I choose **Tool Call** as Fucntion Calling method, the tool message role should be `tool`, otherwise, when **Function Call** is choosen, it should be `function`.

### ❌ Actual Behavior

 Now, Dify does not differentiate the two scenario and uniformly uses `function` as the tool message role.

### PR DESCRIPTION
**Pull request by sunxichen**
# Description

Distinguishing the tool message role in two scenarios: When using Tool Call as the Function Calling method, the message role should be `tool`; whereas, when employing Function Call as the Function Calling method, the message role should be `function`.

Fixes #5622 

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] TODO

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [ ] `optional` I have made corresponding changes to the documentation 
- [ ] `optional` I have added tests that prove my fix is effective or that my feature works
- [ ] `optional` New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by Yeuoly — APPROVED** [review]
LGTM

File: api/core/model_runtime/model_providers/openai_api_compatible/llm/llm.py [hunk 1]
```diff
@@ -636,7 +636,7 @@ def _convert_prompt_message_to_dict(self, message: PromptMessage) -> dict:
             #     "tool_call_id": message.tool_call_id
             # }
             message_dict = {
-                "role": "function",
+                "role": "tool" if credentials and credentials.get('function_calling_type', 'no_call') == 'tool_call' else "function",
```
Comment (Yeuoly) [inline api/core/model_runtime/model_providers/openai_api_compatible/llm/llm.py:639]: On line 687, this method is also called. I think you need to pass the credentials there as well.

**Review by Yeuoly — CHANGES_REQUESTED** [review]

File: api/core/model_runtime/model_providers/openai_api_compatible/llm/llm.py (hunk 1 above)
Comment (sunxichen) [reply api/core/model_runtime/model_providers/openai_api_compatible/llm/llm.py:639]: Thank you for catching that oversight. I have addressed your comment by updating line 687 to include the necessary credentials.

**Review by Yeuoly — APPROVED** [review]
LGTM

**Comment by Yeuoly** [comment]
beautiful contribution, thanks for your contributions!

**Pull request merged**